jax/experimental/ode.py

```diff
@@ -12,343 +12,450 @@
-    """
-    v = np.stack([dy0, dy1, y0, y1, y_mid])
-    a = np.dot(np.hstack([-2 * dt,  2 * dt, np.array([ -8., -8.,  16.])]), v)
-    b = np.dot(np.hstack([ 5 * dt, -3 * dt, np.array([ 18., 14., -32.])]), v)
-    c = np.dot(np.hstack([-4 * dt,      dt, np.array([-11., -5.,  16.])]), v)
-    d = dt * dy0
-    e = y0
-    return a, b, c, d, e
 
+@jax.jit
+def fit_4th_order_polynomial(y0, y1, y_mid, dy0, dy1, dt):
+  """Fit fourth order polynomial over function interval.
+
+  Args:
+      y0: function value at the start of the interval.
+      y1: function value at the end of the interval.
+      y_mid: function value at the mid-point of the interval.
+      dy0: derivative value at the start of the interval.
+      dy1: derivative value at the end of the interval.
+      dt: width of the interval.
+  Returns:
+      Coefficients `[a, b, c, d, e]` for the polynomial
+      p = a * x ** 4 + b * x ** 3 + c * x ** 2 + d * x + e
+  """
+  v = np.stack([dy0, dy1, y0, y1, y_mid])
+  a = np.dot(np.hstack([-2. * dt, 2. * dt, np.array([-8., -8., 16.])]), v)
+  b = np.dot(np.hstack([5. * dt, -3. * dt, np.array([18., 14., -32.])]), v)
+  c = np.dot(np.hstack([-4. * dt, dt, np.array([-11., -5., 16.])]), v)
+  d = dt * dy0
+  e = y0
+  return a, b, c, d, e
+
+
+@functools.partial(jax.jit, static_argnums=(0,))
 def initial_step_size(fun, t0, y0, order, rtol, atol, f0):
-    """Empirically choose initial step size.  Algorithm from:
-    E. Hairer, S. P. Norsett G. Wanner,
-    Solving Ordinary Differential Equations I: Nonstiff Problems, Sec. II.4."""
-    scale = atol + np.abs(y0) * rtol
-    d0 = L2_norm(y0 / scale)
-    d1 = L2_norm(f0 / scale)
-
-    if d0 < 1e-5 or d1 < 1e-5:
-        h0 = 1e-6
-    else:
-        h0 = 0.01 * d0 / d1
-
-    y1 = y0 + h0 * f0
-    f1 = fun(y1, t0 + h0)
-    d2 = (L2_norm(f1 - f0) / scale) / h0
-
-    if d1 <= 1e-15 and d2 <= 1e-15:
-        h1 = np.maximum(1e-6, h0 * 1e-3)
-    else:
-        h1 = (0.01 / np.max(d1 + d2))**(1. / float(order + 1))
-
-    return np.minimum(100 * h0, h1)
-
-
-@partial(jit, static_argnums=(0,))
+  """Empirically choose initial step size.
+
+  Args:
+    fun: Function to evaluate like `func(y, t)` to compute the time
+      derivative of `y`.
+    t0: initial time.
+    y0: initial value for the state.
+    order: order of interpolation
+    rtol: relative local error tolerance for solver.
+    atol: absolute local error tolerance for solver.
+    f0: initial value for the derivative, computed from `func(t0, y0)`.
+  Returns:
+    Initial step size for odeint algorithm.
+
+  Algorithm from:
+  E. Hairer, S. P. Norsett G. Wanner,
+  Solving Ordinary Differential Equations I: Nonstiff Problems, Sec. II.4.
+  """
+  scale = atol + np.abs(y0) * rtol
+  d0 = np.linalg.norm(y0 / scale)
+  d1 = np.linalg.norm(f0 / scale)
+  order_pow = (1. / (order + 1.))
+
+  h0 = jax.lax.cond(np.any(np.asarray([d0 < 1e-5, d1 < 1e-5])),
+                    1e-6,
+                    lambda x: x,
```

Comment:
This looks a little strange to me. The usual reason for using `lax.cond` is to avoid computing the outputs in each branch (hence why `cond` requires the `true_fun`/`false_fun` arguments) but you aren't doing that here.

If I recall, `cond` also has some issues with supporting higher level transformation like `vmap`.

So I think it would probably be better to use the simpler `np.where` here (and everyone else below).